pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,794 @@
+		if updateConditionErr != nil {
+			return setting, fmt.Errorf("update status error %v", updateConditionErr)
+		}
+		return updatedSetting, err
+	}
+
+	logrus.Infof("all pods are stopped")
+	logrus.Infof("all volumes are detached")
+	logrus.Infof("update Longhorn settings")
+	// push LH setting
+	nadName := settingCopy.Annotations[NadStorageNetworkAnnotation]
+	if err = h.updateLonghornStorageNetwork(nadName); err != nil {
+		return setting, fmt.Errorf("Update Longhorn setting error %v", err)
+	}
+	if updatedSetting, err = h.setConfiguredCondition(settingCopy, false, ReasonInProgress, MsgUpdateLonghornSetting); err != nil {
+		return setting, fmt.Errorf("update status error %v", err)
+	}
+
+	return updatedSetting, nil
+}
+
+// calc sha1 hash
+func (h *Handler) sha1(s string) string {
+	hash := sha1.New()
+	hash.Write([]byte(s))
+	sha1sum := hash.Sum(nil)
+	return fmt.Sprintf("%x", sha1sum)
+}
+
+func (h *Handler) checkIsSameHashValue(setting *harvesterv1.Setting) bool {
+	currentHash := h.sha1(setting.Value)
+	savedHash := setting.Annotations[HashStorageNetworkAnnotation]
+	return currentHash == savedHash
+}
+
+func (h *Handler) setHashAnnotations(setting *harvesterv1.Setting) {
+	setting.Annotations[HashStorageNetworkAnnotation] = h.sha1(setting.Value)
+}
+
+func (h *Handler) createAndSaveNad(setting *harvesterv1.Setting) error {
+	var config Config
+	bridgeConfig := NewBridgeConfig()
+
+	if err := json.Unmarshal([]byte(setting.Value), &config); err != nil {
+		return fmt.Errorf("parsing value error %v", err)
+	}
+
+	bridgeConfig.Bridge = config.ClusterNetwork + BridgeSuffix
+	bridgeConfig.IPAM.Range = config.Range
+
+	if config.Vlan == 0 {
+		config.Vlan = DefaultPVID
+	}
+	bridgeConfig.Vlan = config.Vlan
+
+	if len(config.Exclude) > 0 {
+		bridgeConfig.IPAM.Exclude = config.Exclude
+	}
+
+	nadConfig, err := json.Marshal(bridgeConfig)
+	if err != nil {
+		return fmt.Errorf("output json error %v", err)
+	}
+
+	nad := nadv1.NetworkAttachmentDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: StorageNetworkNetAttachDefPrefix,
+			Namespace:    StorageNetworkNetAttachDefNamespace,
+		},
+	}
+	nad.Annotations = map[string]string{
+		StorageNetworkAnnotation: "true",
+	}
+	nad.Spec.Config = string(nadConfig)
+
+	// create nad
+	var nadResult *nadv1.NetworkAttachmentDefinition
+	if nadResult, err = h.networkAttachmentDefinitions.Create(&nad); err != nil {
+		return fmt.Errorf("create net-attach-def failed %v", err)
+	}
+
+	// save nad name
+	setting.Annotations[OldNadStorageNetworkAnnotation] = setting.Annotations[NadStorageNetworkAnnotation]
+	setting.Annotations[NadStorageNetworkAnnotation] = nadResult.Namespace + "/" + nadResult.Name
+
+	return nil
+}
+
+func (h *Handler) checkValueIsChanged(setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	var updatedSetting *harvesterv1.Setting
+	var err error
+	if !h.checkIsSameHashValue(setting) {
+		if setting.Value == "" {
```

Comment:
this will set the default empty value as configured during initialization; we may need to check if both values in the LH and Harvester settings are empty first.

Reply:
I keep this checking point because it will update HARV setting to current format with hash.
So it will test HARV setting with LH setting in the next step.
and do the post config process and set the HARV setting to completed status.

start pod func will be skipped because there is no annotation that needs to restore.

Reply:
Once HARV setting with empty value will act as the same as any future changes.

Reply:
That is also the reason I hope we could have controlled in v1.1.0.
It will be complicated to deal with the initialization case.
and need to set up a special condition for this.

Reply:
what about checking when the value is empty, the LH setting is empty, and the annotation is nil? it should be a valid init case?

Reply:
Ok I will change it.